src/platform/backends/qemu/qemu_mount_handler.cpp

```diff
@@ -38,54 +29,58 @@ namespace
 } // namespace
 
-mp::QemuMountHandler::QemuMountHandler(const SSHKeyProvider& ssh_key_provider) : MountHandler(ssh_key_provider)
+namespace multipass
 {
-}
-
-void mp::QemuMountHandler::init_mount(VirtualMachine* vm, const std::string& target_path, const VMMount& vm_mount)
+QemuMountHandler::QemuMountHandler(QemuVirtualMachine* vm, const SSHKeyProvider* ssh_key_provider,
+                                   const std::string& target, const VMMount& mount)
+    : MountHandler{vm, ssh_key_provider, target, mount},
+      qemu_vm{vm},
+      // Create a reproducible unique mount tag for each mount. The cmd arg can only be 31 bytes long so part of the
+      // uuid must be truncated. First character of tag must also be alpabetical.
+      tag{QUuid::createUuidV3(QUuid(), QString::fromStdString(target))
+              .toString(QUuid::WithoutBraces)
+              .remove("-")
+              .left(30)}
 {
-    using St = VirtualMachine::State;
-    const auto skip_states = {St::off, St::stopped, St::suspended};
-
-    auto state = vm->current_state();
-    if (std::none_of(cbegin(skip_states), cend(skip_states), [&state](const auto& st) { return state == st; }))
-        throw std::runtime_error("Please shutdown virtual machine before defining native mount.");
-
-    if (!MP_FILEOPS.exists(QDir{QString::fromStdString(vm_mount.source_path)}))
+    switch (vm->current_state())
     {
-        throw std::runtime_error(fmt::format("Mount path \"{}\" does not exist.", vm_mount.source_path));
+    case VirtualMachine::State::off:
+    case VirtualMachine::State::stopped:
+    case VirtualMachine::State::suspended:
+        break;
+    default:
+        throw std::runtime_error("Please shutdown the instance before attempting native mounts.");
     }
 
     // Need to ensure no more than one uid/gid map is passed in here.
-    if (vm_mount.uid_mappings.size() > 1 || vm_mount.gid_mappings.size() > 1)
-    {
+    if (mount.uid_mappings.size() > 1 || mount.gid_mappings.size() > 1)
         throw std::runtime_error("Only one mapping per native mount allowed.");
-    }
 
     mpl::log(mpl::Level::info, category,
-             fmt::format("Initializing native mount {} => {} in {}", vm_mount.source_path, target_path, vm->vm_name));
-
-    vm->add_vm_mount(target_path, vm_mount);
-    mounts[vm->vm_name][target_path] = vm;
+             fmt::format("initializing native mount {} => {} in '{}'", mount.source_path, target, vm->vm_name));
+
+    const auto uid_map = mount.uid_mappings.empty() ? std::make_pair(1000, 1000) : mount.uid_mappings[0];
+    const auto gid_map = mount.gid_mappings.empty() ? std::make_pair(1000, 1000) : mount.gid_mappings[0];
+    const auto uid_arg = QString("uid_map=%1:%2,").arg(uid_map.first).arg(uid_map.second == -1 ? 1000 : uid_map.second);
+    const auto gid_arg = QString{"gid_map=%1:%2,"}.arg(gid_map.first).arg(gid_map.second == -1 ? 1000 : gid_map.second);
+    qemu_vm->mount_args[target] = {
```

Comment:
Yeah, I agree that's it's not a nice approach. But we need to set, erase and read from the map, so we need full access to it basically. Would it be nicer to have three separate methods for that, or only one that returns a reference to the map and use that to modify it?

And while we're on this topic, I want to address a related issue and ask for an opinion. The reason we need to read from the map is to help make `stop` idempotent. If we call `stop` once and it succeeds, it will erase the respective entry from the map, and any further call to `stop` will read the map again and not do anything, since there will be no entry in the map. And we do something similar for the other implementations as well. We check some state to determine whether a successful `stop` has happened, and do nothing if it did. Would it be an inelegant approach to just have boolean flags that we set on success and use them to determine whether `start`/`stop` have completed successfully?

Reply:
> [...] we need to set, erase and read from the map, so we need full access to it basically [...]

In such cases, I think it is preferable to have an (appropriately named) accessor method that returns a non-const reference (e.g. `modifiable_mount_args`. That pulls it to the interface explicitly and makes it clear that it can be changed elsewhere. An example of a practical situation this would be helpful for is when grepping the code of QEMU VM class to find where `mount_args` is changed. With the current implementation, it would be easy to miss the friend declaration that enables the change in the mount handler. On the contrary, an accessor would be hit by the grep, so people would know they'd need to search for it too.

> [...] boolean flags [...]

Hmm, good question. On the one hand, that would be a sort of duplication of truth, so there is always the risk of going out of sync. On the other, it would avoid a search. So it could be considered a justified form of caching. Also, I may be missing something, but I think you may need something of that sort in samba mounts because, IIUC, the `share_name` is only set in the constructor and clearing it would prevent the mount from starting again after it was stopped. 

Finally, I guess it makes sense conceptually for the base class that offers stop/start to keep track of whether or not it is active, regardless of actual implementations. So one approach could be to implement start/stop with the *template method* pattern in the base class. That is, the base `MountHandler` would have something like:

```c++
public:
(not virtual) void start(...)
{
    if (!m_active)
    {
        start_impl(...);
        m_active = true;
    }
}

(not virtual) void stop()
{
    if (active)
      stop_impl();
      m_active = true;
}

bool is_active() { return m_active; }

protected:
virtual void start_impl(...) = 0;
virtual void stop_impl() = 0;

private:
bool m_active;
```

Descendants would only implement *_impl versions and not need to care about idempotence any longer &ndash; they could still assert that they were being called from the right state, or even throw a `logic_error` if we fear that situation would arise in real usage. Of course, all of this would also need locks to ensure proper thread-safety...

Reply:
Nice, I like this. Will implement!